File: app/src/main/java/com/example/paritosh/stormy/WeatherPresenterIMPL.java
package com.example.paritosh.stormy;

import com.example.paritosh.stormy.model.CurrentWeather;
import com.example.paritosh.stormy.model.CurrentWeatherDataBindingModel;

public class WeatherPresenterIMPL implements WeatherContract.WeatherPresenter {
    private WeatherContract.WeatherView weatherView;


    WeatherPresenterIMPL(WeatherContract.WeatherView weatherView) {
        this.weatherView = weatherView;

    }

    private WeatherDataProvider weatherDataProvider = new WeatherDataProvider();

    @Override
    public void updateWeatherDetails() {
        if (Utils.isNetworkAvailable(weatherView.getContext())) {
            weatherDataProvider.getCurrentWeather(new WeatherDataProvider.OnWeatherApiResponse() {
                @Override
                public void onSuccess(CurrentWeather weather) {
                    CurrentWeatherDataBindingModel model = Utils.convertCurrentWeatherToDataBindingModel(weather);
                    weatherView.render(model);
                }

                @Override
                public void onError(Throwable t) {
                    weatherView.showApiFailError();
                }
            });
        } else {
            weatherView.showConnectivityError();
        }
    }

    @Override
    public void refreshWeatherData() {
        updateWeatherDetails();
        weatherView.showMessage(R.string.refresh_data_message);
    }


}

File: app/src/main/java/com/example/paritosh/stormy/WeatherPresenterImpl.kt
package com.example.paritosh.stormy

import com.example.paritosh.stormy.model.CurrentWeather

class WeatherPresenterImpl(val view: WeatherContract.WeatherView) : WeatherContract.WeatherPresenter {
    private val weatherDataProvider = WeatherDataProvider()

    override fun updateWeatherDetails() {

        if (Utils.isNetworkAvailable(view.loadContext())) {
            weatherDataProvider.getCurrentWeather(object : WeatherDataProvider.OnWeatherApiResponse {
                override fun onSuccess(weather: CurrentWeather) {
                    val model = Utils.convertWeatherToDataBindingModel(weather)
                    view.render(model)
                }

                override fun onError(t: Throwable) {
                    view.showErrorMessage(R.string.error_title, R.string.error_message)
                }
            })
        } else
            view.showErrorMessage(R.string.error_title, R.string.connectivity_error_message)
    }

    override fun refreshWeatherDetails() {
        updateWeatherDetails()
        view.showMessage(R.string.refresh_data_message)
    }
}